### PR TITLE
Ragged reduction

### PR DESCRIPTION
**Pull request by syurkevi**
Addresses #2782 .

TODO:
* tests
* other ragged functions(min, sum, product, count, any, all)

### CODE REVIEW

File: src/api/c/reduce.cpp [hunk 1]
```diff
@@ -804,6 +820,68 @@ af_err af_imax(af_array *val, af_array *idx, const af_array in, const int dim) {
     return ireduce_common<af_max_t>(val, idx, in, dim);
 }
 
+template<af_op_t op>
+static af_err rreduce_common(af_array *val, af_array *idx, const af_array in,
```
Comment (umar456) [inline src/api/c/reduce.cpp]: Will an overload to reduce_common not work?

File: src/api/c/reduce.cpp [hunk 2]
```diff
@@ -804,6 +820,68 @@ af_err af_imax(af_array *val, af_array *idx, const af_array in, const int dim) {
...
+        ARG_ASSERT(2, in_info.ndims() > 0);
+
+        if (dim >= (int)in_info.ndims()) {
+            *val = retain(in);
+            *idx = createHandleFromValue<uint>(in_info.dims(), 0);
+            return AF_SUCCESS;
+        }
+
+        // TODO: make sure ragged_len.dims == in.dims(), except on reduced dim
+        const ArrayInfo &key_info = getInfo(ragged_len);
```
Comment (umar456) [inline src/api/c/reduce.cpp]:
```suggestion
        const ArrayInfo &ragged_len_info = getInfo(ragged_len);
```

File: src/api/c/reduce.cpp [hunk 3]
```diff
@@ -804,6 +820,68 @@ af_err af_imax(af_array *val, af_array *idx, const af_array in, const int dim) {
...
+        dim4 test_dim = in_info.dims();
```
Comment (umar456) [inline src/api/c/reduce.cpp]:
```suggestion
        dim4 out_dim = in_info.dims();
```

File: include/af/algorithm.h [hunk 4]
```diff
@@ -216,6 +216,23 @@ namespace af
...
+       \param[in] in contains the input values to be reduced
+       \param[in] dim The dimension along which the max operation occurs
+       \param[in] ragged_len The number of elements to look at when reducing along \p dim
+
+       \ingroup reduce_func_max
+
+       \note NaN values are ignored
+    */
+    AFAPI void max(array &val, array &idx, const array &in, const int dim, const array &ragged_len);
+#endif
```
Comment (umar456) [inline include/af/algorithm.h]:
```suggestion
#endif  // AF_API_VERSION >= 38
```

File: src/api/c/reduce.cpp [hunk 5]
```diff
@@ -804,6 +820,68 @@ af_err af_imax(af_array *val, af_array *idx, const af_array in, const int dim) {
...
+        test_dim[dim] = 1;
+        ARG_ASSERT(4, test_dim == key_info.dims());
+        ARG_ASSERT(4, test_dim == key_info.dims());
```
Comment (umar456) [inline src/api/c/reduce.cpp]: These two conditions are the same.

File: src/backend/cpu/ireduce.cpp [hunk 6]
```diff
@@ -24,18 +24,32 @@ namespace cpu {
...
         kernel::ireduce_dim<op, T, 1>(), kernel::ireduce_dim<op, T, 2>(),
         kernel::ireduce_dim<op, T, 3>(), kernel::ireduce_dim<op, T, 4>()};
 
-    getQueue().enqueue(ireduce_funcs[in.ndims() - 1], out, loc, 0, in, 0, dim);
+    getQueue().enqueue(ireduce_funcs[in.ndims() - 1], out, loc, 0, in, 0, dim, rlen);
+}
+
+template<af_op_t op, typename T>
+void rreduce(Array<T> &out, Array<uint> &loc, const Array<T> &in,
+             const int dim, const Array<uint> &rlen) {
```
Comment (umar456) [inline src/backend/cpu/ireduce.cpp]: Is this function still needed? It looks like you can combine this with ireduce.

**Comment by WilliamTambellini** [comment]
@syurkevi  sounds good:
ArrayFire v3.7.0 (CUDA, 64-bit Linux, build 4206fa32)
Platform: CUDA Runtime 10.2, Driver: 440.33.01
[0] Quadro RTX 3000, 5935 MB, CUDA Compute 7.5
Benchmark ragged max at f32 
in		 max 	 ragmax
4	16384	 0.038	 0.037	
8	16384	 0.034	 0.036	
16	16384	 0.035	 0.037	
32	16384	 0.037	 0.040	
64	16384	 0.074	 0.081	
128	16384	 0.148	 0.159	
256	16384	 0.293	 0.309	

Though f16 not really faster than f32:
Device 0 isHalfAvailable ? yes
ArrayFire v3.7.0 (CUDA, 64-bit Linux, build 4206fa32)
Platform: CUDA Runtime 10.2, Driver: 440.33.01
[0] Quadro RTX 3000, 5935 MB, CUDA Compute 7.5
Benchmark ragged max at f16 
in		 max 	 ragmax
4	16384	 0.040	 0.037	
8	16384	 0.035	 0.037	
16	16384	 0.036	 0.039	
32	16384	 0.038	 0.040	
64	16384	 0.067	 0.070	
128	16384	 0.143	 0.152	
256	16384	 0.282	 0.298

Could you please just add a minimalist bench like this one:
<script src="https://gist.github.com/WilliamTambellini/0f9309ab27b7076369f24c3217af4ffd.js"></script>
https://gist.github.com/WilliamTambellini/0f9309ab27b7076369f24c3217af4ffd
?

**Comment by WilliamTambellini** [comment]
@syurkevi  could you please at least rebase for me to see what needs to be finished and advise accordingly ?

**Comment by 9prady9** [comment]
@syurkevi I took care of rebase from latest master. If you are adding more ragged functions and need to touch the ireduce kernel. You can find the kernels in the file `src/backend/cuda/kernel/ireduce.cuh` and the kernel wrappers inside `src/backend/cuda/kernel/ireduce.hpp`. The backend source file is `src/backend/cuda/ireduce.cpp` and not `ireduce.cu`.

If you face any issues while editing/adding new things to kernels, please ping me. I can guide you through the nvrtc related changes.

**Comment by WilliamTambellini** [comment]
Thank you @9prady9 
@syurkevi  I am only interested by the max reduction, as agreed with @umar456 when ordering the support. Could you please limit that PR to the max/raggedmax reduction ?
Kind regards

**Pull request merged**